pkg/controller/networkpolicy/status_controller.go

```diff
@@ -0,0 +1,343 @@
+	internalNetworkPolicyStore storage.Interface
+
+	// statuses is a nested map that keeps the realization statuses reported by antrea-agents.
+	// The outer map's keys are the NetworkPolicy keys. The inner map's keys are the Node names. The inner map's values
+	// are statuses reported by each Node for a NetworkPolicy.
+	statuses     map[string]map[string]*controlplane.NetworkPolicyNodeStatus
+	statusesLock sync.RWMutex
+
+	// cnpLister is able to list/get ClusterNetworkPolicies and is populated by the shared informer passed to
+	// NewClusterNetworkPolicyController.
+	cnpLister seclisters.ClusterNetworkPolicyLister
+	// cnpListerSynced is a function which returns true if the ClusterNetworkPolicies shared informer has been synced at least once.
+	cnpListerSynced cache.InformerSynced
+	// anpLister is able to list/get AntreaNetworkPolicies and is populated by the shared informer passed to
+	// NewNetworkPolicyController.
+	anpLister seclisters.NetworkPolicyLister
+	// anpListerSynced is a function which returns true if the AntreaNetworkPolicies shared informer has been synced at least once.
+	anpListerSynced cache.InformerSynced
+}
+
+func NewStatusController(antreaClient antreaclientset.Interface, internalNetworkPolicyStore storage.Interface, cnpInformer secinformers.ClusterNetworkPolicyInformer, anpInformer secinformers.NetworkPolicyInformer) *StatusController {
+	c := &StatusController{
+		npControlInterface: &networkPolicyControl{
+			antreaClient: antreaClient,
+			anpLister:    anpInformer.Lister(),
+			cnpLister:    cnpInformer.Lister(),
+		},
+		queue:                      workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "networkpolicy"),
+		internalNetworkPolicyStore: internalNetworkPolicyStore,
+		statuses:                   map[string]map[string]*controlplane.NetworkPolicyNodeStatus{},
+		cnpListerSynced:            cnpInformer.Informer().HasSynced,
+		anpListerSynced:            anpInformer.Informer().HasSynced,
+	}
+	// To save a "GET" query before each update, UpdateAntreaClusterNetworkPolicyStatus treats the cache of Lister as
+	// the state of kube-apiserver. However, in some cases the cache may not be in sync, then we might skip updating a
+	// policy's status by mistake. To resolve it, add update event handlers which trigger resync of a policy if its
+	// status is updated.
+	cnpInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.updateCNP,
+		},
+		resyncPeriod,
+	)
+	anpInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.updateANP,
+		},
+		resyncPeriod,
+	)
+	return c
+}
+
+func (c *StatusController) updateCNP(old, cur interface{}) {
+	curCNP := cur.(*secv1alpha1.ClusterNetworkPolicy)
+	oldCNP := old.(*secv1alpha1.ClusterNetworkPolicy)
+	if oldCNP.Status == curCNP.Status {
```

Comment:
Why watch CNP/ANP changes and check their Status here? Is it for the case some other clients change the Status field?

Reply:
It's not designed for the case other clients changing the status field but can work for it.
The main consideration is we need a source of truth of the status so that we can avoid doing dummy CRD update call.
There are two options: get it before updating, which is apparently not worth and doesn't reduce API call number.
Another option is the CRD Lister we already have for networkpolicycontroller. However, Lister is not always in sync with K8s API. For example, if a policy's CurrentNodesRealized is changed from 2 -> 3 -> 2 quickly, in the first round, we want to update it to 3 and found it's 2 in Lister so we will call the update API, in the second round, we want to update it to 2 and found it's also 2 in Lister (it's very possible because there is some delay for the first update to be reflected in Lister) then we won't call the update API, causing the status in API kept 3 wrongly.
Since Lister is our source of truth of status, if it is changed, we need a resync for the status. Also explained it in comments L82-L85.

Reply:
Could we compare with the internal state instead of CRD Status to know if an update is needed?

Reply:
By internal state, do you mean a "desired" status of CR's status? We don't have a separate storage for that and use Lister as the source of truth of data in apiserver. I guess we need to maintain the internal state in the end of syncHandler and use it in update eventhandler to prevent unnecessary tasks enqueued?

I'm not sure if it could solve the above problem: in the second round of above example, if controller hasn't received CR update event from 2 to 3, syncHandler that wants to update CR status from 3 to 2 won't make update API call because CR status in Lister is already 2. If the update event comes before it updates the internal state from 3 to 2, the eventhandler will also skip enqueuing because it thinks the latest CR status 3 matches the internal state 3. or does updating internal state before making API call prevent this problem?

Reply:
I think it is not prevent tasks enqueued, but to prevent unnecessary K8s calls, and we should check the desired status in syncHandler before updating CR status.

But I think you still mean you want to cover the case the status is updated by other clients, so you want to watch CR status change here?

I think the problem is that we are not able to append information to the event (like it is status change, or span change, or generation change), so we probably have a lot unnecessary computation.
But anyway would you add some comments to explain the reasoning of handling status changes (in my mind if not to consider other clients, we can check the last desired state in syncHandler)?

Reply:
> I think it is not prevent tasks enqueued, but to prevent unnecessary K8s calls, and we should check the desired status in syncHandler before updating CR status.
> 
> But I think you still mean you want to cover the case the status is updated by other clients, so you want to watch CR status change here?

It's already done in syncHandler. Yes, it could cover the other clients's update case. I'm curious what do we compare with after restarting controller if not with CR status?

> 
> I think the problem is that we are not able to append information to the event (like it is status change, or span change, or generation change), so we probably have a lot unnecessary computation.

I think this is an old topic, it's doable but more complex and no K8s controller is doing it. We will need to maintain several extra status (not in the workqueue) to optimize syncHandler's process. I believe it requires some storage and more code in both eventHandler and syncHandler, unless the function becomes a bottleneck I think it may be not worth.

> But anyway would you add some comments to explain the reasoning of handling status changes (in my mind if not to consider other clients, we can check the last desired state in syncHandler)?

Does the comment in https://github.com/vmware-tanzu/antrea/pull/1442/files#diff-372db8becab5487b98d6104ebfb62f30288bbe258cfe9f2da67333c6cc4f4b22R82 work for you? We don't have storage for last desired state, and even we make one, we need something to compare for restart case when last desired state will be empty.

Reply:
I think my point about internal status vs CR status is we already have the latter. The former will be kind of redundant with it and cannot solve the restart case and the other client's updates case.

Reply:
I just do not like the fact very CR status update will trigger a new event -> status recomputation. In a sense we double the computation?

Reply:
You are correct that it needs to compute it another time after a successful update. This is also an inefficient part of most K8s controllers.
I could add a storage to maintain the expected status of CR, then use it in eventHandler, only enqueue the policy when the new status doesn't match the expected status, what do you think?

Reply:
How about let us merge the current implementation first, then think about optimization?
I just hope you can describe these design considerations for us (probably mainly me as you know the code well) to track and understand later.

Reply:
Sure, added more comments and TODO to L82-L88 to follow up.